rir/src/ir/Compiler.cpp

```diff
@@ -35,33 +36,89 @@ class Context {
       public:
         CodeStream cs;
         std::stack<LoopContext> loops;
-        CodeContext(SEXP ast, FunctionHandle& fun) : cs(fun, ast) {}
+        CodeContext* parent;
+        CodeContext(SEXP ast, FunctionHandle& fun, CodeContext* p) : cs(fun, ast), parent(p) {}
+        virtual ~CodeContext() {}
+        bool inLoop() {
+            return !loops.empty() ||
+                    (parent && parent->inLoop());
+        }
+        LabelT loopNext() {
+            return !loops.empty() ? loops.top().next_ : parent->loopNext();
+        }
+        LabelT loopBreak() {
+            return !loops.empty() ? loops.top().break_ : parent->loopBreak();
```

Comment:
I think those two are wrong. Wouldn't this cause the next inside the promise to use the label of the outer function which is not a valid label in the promise. (The bug probably did not happen in practice since gotoOK returns false in that case.)